File: T08_37_15594_Amr_Kayid.py
# T08_37_15594_Amr_Kayid
__author__ = "Amr Kayid"

import random
from pprint import pformat
from collections import namedtuple, OrderedDict

rl = namedtuple('rule', ['left', 'right'])


class CFG:
  """
  For this task you will implement the context-free grammar (CFG)
  left-recursion elimination algorithm introduced in Lecture 3 of CSEN1003.
  Recall that a CFG is a quadruple (V, Σ, R, S) where
  V and Σ are disjoint alphabets (respectively, containing variables and terminals),
  R ⊆ V ×(V ∪Σ)∗ is a set of rules,
  and S∈V is the start variable.

  """

  def __init__(self, description: str = ""):
    self.start_variable = 'S'
    self.variables = set(filter(str.isupper, description))
    self.terminals = set(filter(str.islower, description))
    self.desc = description.strip().split(';')
    self.rules = OrderedDict({
        i: v for i, v in enumerate([
            rl(
                left=rule.split(',')[0],
                right=tuple(rule.split(',')[1:]),
            ) for rule in self.desc
        ])
    })

    # print(pformat(vars(self)))

  def step(self, rule):
    lhs, rhs = rule.left, rule.right
    alphas = (t[1:] for t in rhs if t[0] == lhs)
    betas = (t for t in rhs if not (t[0] == lhs))
    lhs_new = lhs + '\''
    old_rule = rl(lhs, right=tuple([b + lhs_new for b in betas]))
    new_rule = rl(lhs_new, right=tuple([a + lhs_new for a in alphas] + ['']))
    return old_rule, new_rule

  def is_left_rec(self, rule):
    return any(rule.left == t[0] for t in rule.right)

  def con_2(self, rule):
    return any(v == t[0] for t in rule.right for v in self.variables)

  def step2(self, rule):
    new_rule = None
    for r in self.rules.values():
      l = []
      for t in rule.right:
        if r.left == t[0]:
          l.append([rt + t[1:] for rt in r.right])
        else:
          l.append([t])
      if len(l) > 0:
        rule = rl(rule.left, tuple(sum(l, [])))
      if self.is_left_rec(rule):
        rule, new_rule = self.step(rule)
      # print(rule, new_rule)
    return rule, new_rule

  def _run(self):
    i, new_cfg = 0, OrderedDict()
    for index, rule in self.rules.items():
      if self.is_left_rec(rule):
        rule_, new_rule = self.step(rule)
        self.rules[index] = rule_
        new_cfg[i], new_cfg[i + 1], i = rule_, new_rule, i + 2
      elif self.con_2(rule) and not (index == 0):
        new_cfg[i], new_rule = self.step2(rule)
        i += 1
        if not new_rule == None:
          new_cfg[i] = new_rule
          i += 1
      else:
        new_cfg[i] = rule
        i += 1

    return new_cfg

  def lre(self):
    output = self._run()
    out_str = "".join([
        f'{rule.left},{",".join(t for t in rule.right)};'
        for rule in output.values()
    ])
    return out_str


if __name__ == '__main__':
  inputs = [
      "S,ScT,T;T,aSb,iaLb,i;L,SdL,S",
      "S,Sa,b",
      "S,Sab,cd",
      "S,SuS,SS,Ss,lSr,a",
      "S,SuT,T;T,TF,F;F,Fs,P;P,a,b",
      "S,z,To;T,o,Sz",
      "S,BC,C;B,Bb,b;C,SC,a",
  ]
  outputs = [
      "S,TS';S',cTS',;T,aSb,iaLb,i;L,aSbS'dL,iaLbS'dL,iS'dL,aSbS',iaLbS',iS';",
      "S,bS';S',aS',;",
      "S,cdS';S',abS',;",
      "S,lSrS',aS';S',uSS',SS',sS',;",
      "S,TS';S',uTS',;T,FT';T',FT',;F,PF';F',sF',;P,a,b;",
      "S,z,To;T,oT',zzT';T',ozT',;",
      "S,BC,C;B,bB';B',bB',;C,bB'CCC',aC';C',CC',;",
  ]
  for i, s in enumerate(inputs):
    cfg = CFG(s)
    out = cfg.lre()
    if (out == outputs[i]):
      print(f"[{True}] String: {s:>30} ==> Output: {out}")
    else:
      print(f"[!{True}] String: {s:>30} ==> Output: {out}")

File: T08_37_15594_Amr_Kayid.java
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/***
 * Author: Amr Kayid
 * ID: 37-15594
 * Tut: T08
 * T08_37_15594_Amr_Kayid
 */

public class T08_37_15594_Amr_Kayid {
	public static void main(String[] args) {
		String input = "S,ScT,T;T,aSb,iaLb,e;L,SdL,S";
		CFG cfg = new CFG(input);
		System.out.println(cfg);
		String firstEncoding = cfg.First();
		String followEncoding = cfg.Follow();
		System.out.println("First: " + firstEncoding);
		System.out.println("Follow: " + followEncoding);
	}

	/***
	 * For this task you will implement the algorithms computing the functions First and Follow,
	 * introduced in Lecture 4 of CSEN1003, for the variables of a given context-free grammar.
	 */

	static class CFG {
		/**
		 * Recall that a CFG is a quadruple (V, Σ, R, S)
		 * where V and Σ are disjoint alphabets (respectively, containing variables and terminals),
		 * R ⊆ V × (V ∪ Σ)∗ is a set of rules,
		 * and S ∈ V is the start variable.
		 * >>> Input Example:
		 * "S,ScT,T;T,aSb,iaLb,e;L,SdL,S"
		 */

		public String start = "S";
		public Set<String> vars = new HashSet<>();
		public Set<String> terminals = new HashSet<>();
		public Map<String, ArrayList<String>> rules = new LinkedHashMap<>();

		public Map<String, Set<String>> first = new LinkedHashMap<>();
		public Map<String, Set<String>> follow = new LinkedHashMap<>();
		public Set<String> eps = new HashSet<>(Collections.emptyList());

		public CFG(String input) {
			input.chars().filter(c -> Character.isUpperCase(c)).forEach(v -> vars.add((char) v + ""));
			input.chars().filter(c -> Character.isLowerCase(c)).forEach(t -> terminals.add((char) t + ""));

			terminals.remove("e");

			String[] r = input.split(";");
			for (String rule : r) {
				String[] split = rule.split(",");
				rules.put(split[0], new ArrayList(Arrays.asList(Arrays.copyOfRange(split, 1, split.length))));
			}
			rules.entrySet().forEach(stringArrayListEntry -> {
				if (stringArrayListEntry.getValue().contains("e")) {
					stringArrayListEntry.getValue().add("");
					stringArrayListEntry.getValue().remove("e");
				}
			});


			ComputeFirstFollow(this);
		}

		public static void ComputeFirstFollow(CFG g) {
			Map<String, Set<String>> first = new LinkedHashMap<>();
			g.vars.forEach(v -> first.put(v, new HashSet<>(Collections.emptyList())));
			g.terminals.forEach(t -> first.put(t, new HashSet<>(Collections.singletonList(t))));

			Map<String, Set<String>> follow = new LinkedHashMap<>();
			g.vars.forEach(v -> follow.put(v, new HashSet<>(Collections.emptyList())));

			Set<String> EPSILON = new HashSet<>(Collections.emptyList());

			boolean updated;
			while (true) {
				updated = false;

				for (Map.Entry<String, ArrayList<String>> entry : g.rules.entrySet()) {
					String nt = entry.getKey();
					for (String expression : entry.getValue()) {

						// first
						for (char symbol : expression.toCharArray()) {
							updated |= union(first.get(nt), first.get(symbol + ""));
							if (!EPSILON.contains(symbol + ""))
								break;
						}
						updated |= union(EPSILON, new HashSet<>(Collections.singleton(nt)));

						// follow
						Set<String> aux = follow.get(nt);
						for (Character symbol : new StringBuilder(expression).reverse().toString().toCharArray()) {
							if (follow.containsKey(symbol + "")) {
								updated |= union(follow.get(symbol + ""), aux);
							}
							if (EPSILON.contains(symbol + "")) {
								aux = Stream.concat(aux.stream(), first.get(symbol + "").stream())
										.collect(Collectors.toSet());
							} else {
								aux = first.get(symbol + "");
							}
						}
					}
				}

				if (!updated) {
					g.first = first;
					g.follow = follow;
					g.eps = EPSILON;
					return;
				}
			}

		}

		public static boolean union(Set<String> first, Set<String> begins) {
			int n = first.size();
			first.addAll(begins);

			return first.size() != n;
		}

		public String First() {
			String output = "";
			eps.forEach(v -> {
				first.get(v).add("e");
			});
			return getString(output, first);
		}

		public String Follow() {
			String output = "";
			follow.get(start).add("$");
			rules.get(start).forEach(s -> {
				for (char symbol : s.toCharArray()) {
					if (vars.contains(symbol + ""))
						follow.get(symbol + "").add("$");
				}
			});
			return getString(output, follow);
		}

		private String getString(String output, Map<String, Set<String>> string) {
			output += start + "," + String.join("", string.get(start)) + ";";
			for (Map.Entry<String, Set<String>> entry : string.entrySet()) {
				String key = entry.getKey();
				if (vars.contains(key) && !key.equals(start)) {
					String value = String.join("", entry.getValue());
					if (value.contains("$")) {
						value = value.replace("$", "");
						value += "$";
					}
					output += key + "," + value + ";";
				}
			}
			return output.substring(0, output.length() - 1);
		}

		@Override
		public String toString() {
			return "CFG{" +
					"start='" + start + '\'' +
					", vars=" + vars +
					", terminals=" + terminals +
					", rules=" + rules +
					", first=" + first +
					", follow=" + follow +
					", eps=" + eps +
					'}';
		}
	}
}
